src/coreclr/jit/decomposelongs.cpp

```diff
@@ -137,7 +137,12 @@ GenTree* DecomposeLongs::DecomposeNode(GenTree* tree)
         }
     }
 
+#if defined(FEATURE_HW_INTRINSICS) && defined(TARGET_X86)
+    if (!tree->TypeIs(TYP_LONG) &&
+        !(tree->OperIs(GT_CAST) && varTypeIsLong(tree->AsCast()->CastOp()) && varTypeIsFloating(tree)))
```

Comment:
nit: negated conditions like this can be hard to read. A small comment covering that we want to handle nodes that produce `long` or `GT_CAST float->long` would be beneficial IMO.